### PR TITLE
Db refactor

### LINKED ISSUE #7: Refactor: Database Schema
This requires tweaking the database to migrated to one that has better typing/formatting rules. Right now, it's kind of a mess, and even worse if we ever have to upgrade the server.

Look into ways to potentially make this easier to work with that can also build queries. There are a few python libraries that can help.

### PR DESCRIPTION
**Pull request by user4752**
closes #7 

An opinionated database refactor to convert current raw SQL queries to an ORM library (SQLAlchemy) in order for the data returned from the database to be agnostic to the database engine selected, and allow for clarity on which fields are referenced, i.e. object.named_value versus object[random_index].

Will also allow for easier (and automatic) future migrations of the database should the project needs grow, or the database engine of choice change, e.g. sqlite to mysql/mariadb or postgres.

Opinions chosen are:

- all fields are now properly typed for safety with sane defaults to allow cleaner object creation in code.
- all fields are now named to more clearly identify the values to be stored in them should an external editor be used for raw access
- all data entered into the database now follows the soft-delete paradigm, allowing for a basic historical record of the data, i.e. no data is truly deleted, just marked as 'should not be shown in results unless explicitly specified'
- all data contains basic historical record of last value updated timestamp for possible audit purposes.
- all fields referencing a discord value have been set to width of 32, as maximum discord id numbers are 18 in length, and maximum textual identifiers (nicknames) are 32 in length.
- fields using datetime stamps are using the database internal function to generate the timestamp for uniformity; thus all timestamps are in UTC ensuring no timezone 'drift' which could cause confusion. This also allows for displaying time in either server location (hardcoded) / client location (possibly dynamic) timezone without need to worry about the source timezone due to said uniformity.
- caveat to above is an internal datetime which is used by the softdelete extension of the orm library, it forces use of python's datetime function, but thankfully in utc.

Visualization of the new database schema:
Screenshot 2023-10-14 225122

Examples of the new database objects:
Screenshot 2023-10-14 225132
Screenshot 2023-10-14 225138

Todo:

- refactor logic checks which are spread and varying between the calling function, the internals of called function, and the processing of returned data back in the called function to be uniform. and preferably limited to the database class for safety.
- refactor returned datasets to be proper immutable iterator objects (via list() on return) versus returning the iterator object of the database result which can only be iterated once.
- devise a way to allow backups to be done with other database engines to grow the project's options moving forward

Current Caveats:

- do not have a discord to validate all functionality, thus testing has been limited to all functions being tested in isolation; however all code has been confirmed to execute properly and data modification has been validated.

### CODE REVIEW

**Comment by SocksTheWolf** [comment]
Just reading the opinions of the setup here, I'll look through the code either tonight or tomorrow.

While I agree with a lot of the direction here, I am not sure about the soft deletion practice (specifically around servers). Discord data policies require that if a bot is no longer in a server:

- Information about that server is no longer gathered (does not apply to us)
- Any data stored about that server is removed in a timely manner

I couldn't find what is deemed timely, so this cleanup process could just be whenever the bot next does a reconcile operation, or have it on a timer (we can push this in outside of this pr).

**Comment by user4752** [comment]
> While I agree with a lot of the direction here, I am not sure about the soft deletion practice (specifically around servers). Discord data policies require that if a bot is no longer in a server:
> 
> * Information about that server is no longer gathered (does not apply to us)
> * Any data stored about that server is removed in a timely manner

The soft deletion was meant more as a simplified audit log saying "oh we've interacted with server at least once before" / "we've unbanned this user once before", but it can easy be removed, or actually migrated to a proper audit log. Declaring it as a proper audit log may allow retention to better fall under the 'Applications stated functionality clause' (§5.b ¶2).

Either way, removal is trivial, and would remove one dependency if done so.

**Comment by SocksTheWolf** [comment]
Okay I've been super busy these last few days so I haven't had a chance to look over stuff. I will try to get some time in early next week at the latest.

I'll try to start the review today.

File: BotDatabase.py [hunk 1]
```diff
@@ -215,34 +290,45 @@ def RemoveBan(self, TargetId:int):
...
     def GetAllBans(self, NumLastActions:int=0):
-        LimitStr:str = ""
-        if (NumLastActions > 0):
-            LimitStr = f" LIMIT {NumLastActions}"
-        BansListQuery = self.Database.execute(f"SELECT Id, BannerName FROM banslist ORDER BY ROWID DESC{LimitStr}")
-        return BansListQuery.fetchall()
+        stmt = select(Ban).order_by(desc(Ban.created_at))
+        if (NumLastActions):
+            stmt = stmt.limit(NumLastActions)
+        return self.Database.execute(stmt).scalars(stmt)
```
Comment (SocksTheWolf) [inline BotDatabase.py:325]: Can this return a list wrap so we don't have to make the change in every file?

File: BotDatabase.py [hunk 2]
```diff
@@ -82,47 +105,81 @@ def CleanupBackups(self):
...
+        result = self.Database.scalars(stmt).first()
+
+        if (result is None):
+            Logger.Log(LogLevel.Warn, f"Bot had attempted to set new owner on non-existant server ")
+            return False
+        
+        result.discord_owner_user_id = NewOwnerId
+        self.Database.add(result)
         self.Database.commit()
+        return
```
Comment (SocksTheWolf) [inline BotDatabase.py:137]: Remove extra return here

File: BotDatabase.py [hunk 3]
```diff
@@ -19,16 +26,28 @@ def __del__(self):
 
     def Open(self):
         self.Close()
-        self.Database = sqlite3.connect(Config.GetDBFile())
+        ConfigData=Config()
```
Comment (SocksTheWolf) [inline BotDatabase.py:29]: There's no need to initialize this, bringing in from the import, means that you can just use the static methods easily.

File: BotDatabase.py [hunk 4]
```diff
@@ -19,16 +26,28 @@ def __del__(self):
...
             self.Database = None
             
     def IsConnected(self) -> bool:
-        if (self.Database is not None):
-            return True
+        if (self.Database is None):
+            return False
+
+        if (self.Database.get_bind().closed):
+                return True
```
Comment (SocksTheWolf) [inline BotDatabase.py:50]: Fix excess tab

File: BotDatabase.py [hunk 5]
```diff
@@ -82,47 +105,81 @@ def CleanupBackups(self):
...
-            self.Database.execute(f"DELETE FROM servers where Id={ServerId}")
-            self.Database.commit()
-        else:
-            Logger.Log(LogLevel.Warn, f"Attempted to remove server {ServerId} but we are not in that list!")
+        stmt = select(Server).where(Server.discord_server_id==ServerId).execution_options(include_deleted=True)
+        result = self.Database.scalars(stmt).first()
+
+        if (result is None):
+            Logger.Log(LogLevel.Warn, f"Attempted to remove an invalid server id: {ServerId}")
+            return False
```
Comment (SocksTheWolf) [inline BotDatabase.py:146]: Is this method supposed to return a value?

File: BotDatabase.py [hunk 6]
```diff
@@ -82,47 +105,81 @@ def CleanupBackups(self):
...
+        
+        result.delete()
+        self.Database.commit()
+        return
```
Comment (SocksTheWolf) [inline BotDatabase.py:150]: Remove excess return

File: BotSetup.py [hunk 7]
```diff
@@ -1,30 +1,115 @@
...
-    cursor.execute("CREATE TABLE if not EXISTS banslist(Id, BannerName, BannerId, Date)")
-    cursor.execute("CREATE TABLE if not EXISTS servers(Id, OwnerId, Activated)")
-    con.commit()
-    con.close()
-    Logger.Log(LogLevel.Notice, "Created the bot databases!")
-
-SetupDatabases()
\ No newline at end of file
+
+    ConfigData=Config()
```
Comment (SocksTheWolf) [inline BotSetup.py:17]: This doesn't need to be initialized here, the functions being called are static methods.

File: BotDatabase.py [hunk 8]
```diff
@@ -82,47 +105,81 @@ def CleanupBackups(self):
...
     def SetNewServerOwner(self, ServerId:int, NewOwnerId:int):
-        ActivationChanges = []
-        ActivationChanges.append({"Id": ServerId, "OwnerId": NewOwnerId})
-        self.Database.executemany("UPDATE servers SET OwnerId=:OwnerId WHERE Id=:Id", ActivationChanges)
+        stmt = select(Server).where(Server.discord_server_id==ServerId)
+        result = self.Database.scalars(stmt).first()
+
+        if (result is None):
+            Logger.Log(LogLevel.Warn, f"Bot had attempted to set new owner on non-existant server ")
+            return False
```
Comment (SocksTheWolf) [inline BotDatabase.py:132]: Should this be returning a value?

File: BotDatabase.py [hunk 9]
```diff
@@ -82,47 +105,81 @@ def CleanupBackups(self):
...
         
     def SetNewServerOwner(self, ServerId:int, NewOwnerId:int):
-        ActivationChanges = []
-        ActivationChanges.append({"Id": ServerId, "OwnerId": NewOwnerId})
-        self.Database.executemany("UPDATE servers SET OwnerId=:OwnerId WHERE Id=:Id", ActivationChanges)
+        stmt = select(Server).where(Server.discord_server_id==ServerId)
+        result = self.Database.scalars(stmt).first()
+
+        if (result is None):
+            Logger.Log(LogLevel.Warn, f"Bot had attempted to set new owner on non-existant server ")
```
Comment (SocksTheWolf) [inline BotDatabase.py:131]: Print out the ServerId

File: BotDatabase.py [hunk 10]
```diff
@@ -159,39 +216,46 @@ def ReconcileServers(self, Servers):
...
+
+        return True
     
     # Returns the banner's name, the id and the date
     def GetBanInfo(self, TargetId:int):
-        return self.Database.execute(f"SELECT BannerName, BannerId, Date FROM banslist WHERE Id={TargetId}").fetchone()
+        stmt = select(Ban).where(Ban.discord_user_id==TargetId)
+        result = self.Database.execute(stmt).scalars(stmt).first()
+        if (result is None):
+            return False
```
Comment (SocksTheWolf) [inline BotDatabase.py:256]: This should just return None, so this function can simply just `return result`

**Review by SocksTheWolf — CHANGES_REQUESTED** [review]
Thank you very much for this, I apologize for taking so long to get to it. Had a couple unforeseen circumstances come up, so I was unable to review it

**Comment by user4752** [comment]
> Thank you very much for this, I apologize for taking so long to get to it. Had a couple unforeseen circumstances come up, so I was unable to review it until now.

Not a worry, I was aware of the craziness of last week and can only imagine the stress it caused.

Most of your comment replies are on my aforementioned todo list; this initial set of commits was to gauge interest, design and compatibility. Will be commenting directly back on the individual issues shortly.

**Comment by user4752** [comment]
Was there a final opinion on the soft-delete, and/or an opinion on the reported history / audit log (ref #36)? The former to remove if need be, and the later to prep for implementation in the schema as it will directory dictate some of my replies to the review.

File: BotDatabase.py (hunk 1 above)
Comment (user4752) [reply BotDatabase.py:325]: Yes, will convert all returns to list()s

File: BotDatabase.py [hunk 11]
```diff
@@ -215,34 +290,45 @@ def RemoveBan(self, TargetId:int):
...
-        ActivatedFilter:str = ""
-        if (ActivatedOnly):
-            ActivatedFilter = " WHERE Activated=1"
-        AllServersQuery = self.Database.execute(f"SELECT Id, OwnerId, Activated FROM servers{ActivatedFilter}")
-        return AllServersQuery.fetchall()
+    def GetAllServers(self, ActivationState:bool=False):
+        stmt = select(Server)
+        if (ActivationState):
+            stmt = stmt.where(Server.activation_state==ActivationState)
+        return self.Database.execute(stmt).scalars(stmt)
```
[the remark this answers is not in the document]
Comment (user4752) [reply BotDatabase.py:331]: Yes, will convert all returns to list()s

File: BotDatabase.py [hunk 12]
```diff
@@ -215,34 +290,45 @@ def RemoveBan(self, TargetId:int):
...
+        return result
     
     def GetAllBans(self, NumLastActions:int=0):
-        LimitStr:str = ""
-        if (NumLastActions > 0):
-            LimitStr = f" LIMIT {NumLastActions}"
-        BansListQuery = self.Database.execute(f"SELECT Id, BannerName FROM banslist ORDER BY ROWID DESC{LimitStr}")
-        return BansListQuery.fetchall()
+        stmt = select(Ban).order_by(desc(Ban.created_at))
+        if (NumLastActions):
```
[the remark this answers is not in the document]
Comment (user4752) [reply BotDatabase.py:323]: Will correct

File: BotDatabase.py (hunk 2 above)
Comment (user4752) [reply BotDatabase.py:137]: Was unaware of python's automatic explicit return value on end of definition. Will remove

File: BotDatabase.py (hunk 3 above)
Comment (user4752) [reply BotDatabase.py:29]: Artifact from code testing in isolation. Will remove

File: BotDatabase.py (hunk 6 above)
Comment (user4752) [reply BotDatabase.py:150]: Was unaware of python's automatic explicit return value on end of definition. Will remove

File: BotSetup.py (hunk 7 above)
Comment (user4752) [reply BotSetup.py:17]: Will remove

File: BotDatabase.py (hunk 9 above)
Comment (user4752) [reply BotDatabase.py:131]: Will amend

**Comment by SocksTheWolf** [comment]
> Was there a final opinion on the soft-delete, and/or an opinion on the reported history / audit log (ref #36)? The former to remove if need be, and the later to prep for implementation in the schema as it will directory dictate some of my replies to the review.

Yes, I don't think it's something we should pursue.

**Comment by user4752** [comment]
Closing pull request due to conflicts with v1.5.0 release. Will rebase and open new pull request soon.

**Pull request closed without merging**